src/OPTokenClaim.sol

```diff
@@ -54,62 +70,139 @@ contract OPTokenClaim is Ownable {
-            !epochToAddressClaimed[epoch][account],
-            "already claimed for this epoch"
-        );
+        // users claim reward for last epoch, so claims start at epoch 1
+        uint256 epochNum = _currentEpoch(_config);
+        require(epochNum > 0, "claims have not started yet");
 
-        uint256 claimableOP = _calcReward(account);
+        uint256 lastEpochNum;
+        unchecked {
+            lastEpochNum = epochNum - 1;
+        }
+        require(lastEpochNum < _config.maxEpoch, "claims ended");
+
+        // check if subscribed and if already claimed
+        require(epochToSubscribedEXP[lastEpochNum][account] > 0, "didn't subscribe or already claimed");
+
+        uint256 OPReward = _calcReward(account, lastEpochNum);
 
         // set claimed to true
-        epochToAddressClaimed[epoch][account] = true;
+        epochToSubscribedEXP[lastEpochNum][account] = 0;
 
         // transfer OP to account
-        require(
-            OP.transferFrom(treasury, account, claimableOP),
-            "Transfer failed"
-        );
+        require(OP.transferFrom(treasury, account, OPReward), "Transfer failed");
+
+        emit OPClaimed(account, lastEpochNum, OPReward);
 
-        emit OPClaimed(account, epoch, claimableOP);
+        if (epochNum < _config.maxEpoch) {
+            // subscribe for next epoch
+            subscribe(account);
+        }
     }
 
     /// @dev reverts if claims have not started yet
-    function currentEpoch(Config memory _config)
-        internal
-        view
-        returns (uint256 epochNumber)
-    {
-        require(block.timestamp >= _config.start, "claim period not started");
+    function _currentEpoch(Config memory _config) internal returns (uint256 epochNumber) {
+        require(block.timestamp >= _config.start, "reward dist not started yet");
         unchecked {
             epochNumber = (block.timestamp - _config.start) / 30 days;
         }
+
+        // push new epoch to array if needed
+        while ((epochs.length - 1) < epochNumber) {
+            epochs.push();
+        }
     }
 
-    function _calcReward(address account) internal view returns (uint256) {
-        uint256 expBalance = EXP.balanceOf(account);
-        require(expBalance > 0, "address has no exp");
+    // calculates reward of account for given epoch
+    function _calcReward(address account, uint256 epoch) internal view returns (uint256) {
+        // calculate the total reward of given epoch
+        uint256 totalEXP = epochs[epoch].totalEXP;
+        uint256 subscribedAccounts = epochs[epoch].numAccounts;
 
-        // cap at 99 EXP
-        if (expBalance > 99 ether) {
-            expBalance = 99 ether;
-        }
+        unchecked {
+            uint256 totalReward = totalEXP * 5 - 4 ether * subscribedAccounts;
 
-        return (expBalance * 5 - 4 ether);
-    }
+            // if total reward of given epoch is greater than 10k OP (MAX_REWARD), reduce reward
+            uint256 factor = 1 * 10 ** 18;
```

Comment:
maybe use 1 ether instead of 1 * 10**18?

Reply:
simplified in 2d3841682d14ebc5ca8a33429924c3e9c2ab6d86